src/form/form.js

```diff
@@ -0,0 +1,669 @@
+                                /**
+                                 * @event widgetadd
+                                 * @param {String} uri
+                                 * @param {widgetForm} widget
+                                 */
+                                this.trigger('widgetadd', definition.uri, widget);
+
+                                resolve(widget);
+                            });
+                    });
+                });
+        },
+
+        /**
+         * Removes a widget
+         * @param {String} uri
+         * @returns {form}
+         * @fires widgetremove after the widget has been removed
+         */
+        removeWidget(uri) {
+            if (widgets.has(uri)) {
+                removeComponent(widgets, uri);
+
+                /**
+                 * @event widgetremove
+                 * @param {String} uri
+                 */
+                this.trigger('widgetremove', uri);
+            }
+            return this;
+        },
+
+        /**
+         * Gets the list of widgets.
+         * @returns {Object}
+         */
+        getWidgets() {
+            return getComponents(widgets);
+        },
+
+        /**
+         * Replace the widgets
+         * @param {widgetConfig[]} definitions
+         * @returns {Promise<widgetForm[]>}
+         */
+        setWidgets(definitions) {
+            this.removeWidgets();
+            return Promise.all(_.map(definitions, definition => this.addWidget(definition)));
+        },
+
+        /**
+         * Removes all widgets
+         * @returns {form}
+         */
+        removeWidgets() {
+            for (let uri of widgets.keys()) {
+                this.removeWidget(uri);
+            }
+            widgets.clear();
+            return this;
+        },
+
+        /**
+         * Gets a button by its identifier
+         * @param {String} id
+         * @returns {button}
+         */
+        getButton(id) {
+            return buttons.get(id) || null;
+        },
+
+        /**
+         * Adds a button to the form
+         * @param {buttonConfig} definition
+         * @returns {Promise<button>}
+         * @throws TypeError if the button definition is invalid
+         * @fires button when the button is triggered
+         * @fires button-<id> when the button is triggered
+         * @fires buttonadd after the button has been added
+         */
+        addButton(definition) {
+            return validateDefinition(this, definition, 'id')
+                .then(() => new Promise(resolve => {
+                    const button = buttonFactory(definition);
+                    buttons.set(definition.id, button);
+                    button
+                        .on('click.form', () => {
+                            /**
+                             * @event button
+                             * @param {String} id
+                             */
+                            this.trigger('button', definition.id);
```

Comment:
Is it possible to not trigger an event called `button` but directly the unlerying action : `publish`, `submit`, `reset` or `clear` ?

Reply:
or what about a definition that contains the action : 
```js
return formFactory(container, {
  buttons : [{
        type: 'info',
        id: 'submit',
        label: config.submitText,
        icon: config.submitIcon,
        action : form => form.submit()
), {
      type: 'neutral',
      id: 'reset',
      label: config.resetText,
      icon: config.resetIcon
      action : form => form.reset()
), {
      type: 'foo',
      id: 'reset',
      label: config.resetText,
      icon: config.resetIcon
      action : form => {
          form.trigger('foo')
      }
}])
```

Reply:
Well the idea was to avoid conflict with other events, hence the namespace.
So to listen to any button click, the `button` event is available. For a particular button, the event `button-<id of the button>` is available.
Having action built with the button definition could be also an option, but it will reduce the possibilities, as only one listener will be applicable, and no event pipeline will be available too.